### DISCUSSION TITLE
how to use code-server connect to my own code-marketplace server in my private network(not connect Internet)

### DISCUSSION THREAD

**Discussion by jeffliao888** [coder/code-server#6206] [Q&A] [answered]
Hi, I have setup a code-marketplace server(using a TLS-terminating reverse proxy with NGINX) in my private network already, and I have Open a brower for testing https://myserver.domain.com/ , and then I can see the Output “Marketplace is running” 

I add some extensions to the code-marketplace server with the following command:
./code-marketplace add https://open-vsx.org/api/ms-python/python/2022.14.0/file/ms-python.python-2022.14.0.vsix --extensions-dir ./extensions

then I use the following command to make a query and check one of the URLs in the response:
>curl -k 'https://test888.com/api/extensionquery' -H 'Accept: application/json;api-version=3.0-preview.1' --compressed -H 'Content-Type: application/json' --data-raw '{"filters":[{"criteria":[{"filterType":8,"value":"Microsoft.VisualStudio.Code"}],"pageSize":1}],"flags":439}' | jq

I get some responses from the query so I think the code-marketplace server is worked OK.

then I run the following command(CentOS 7) to up the  code-server:
> export EXTENSIONS_GALLERY='{"serviceUrl":"https://test888.com"}'
> systemctl enable --now code-server@jeff

after that, I use the code server with my brower http://localhost:8080 , and then it seems can't connect to my code-marketplace server to get the extensions, when I click the Extensions , it show error messages which is "Error while fetching extensions, getaddrinfo ENOTFFOUND open-vsx.org" 
I think code-server still can't connect to my own code-marketplace server.

I have no idea how to fix it if anyone can give me a hand.

Thank you very much !

**Comment by code-asher** [comment]
Since you are running it with systemd you will need to set the environment variable in the systemd service file (it will not read the variables in your current terminal session).

Under the `[Service]` block you can add `Environment="EXTENSIONS_GALLERY=<json>"`.

You could also put the environment variable in a file and use `EnvironmentFile` in the service file.

**Reply by jeffliao888** [reply]
Dear @code-asher , thank you for your help.

#Method 1 : I put the environment variable in a file and use EnvironmentFile in the service file as you said.
I add one line under the [Service] block
> EnvironmentFile=/etc/myenv

and the /etc/myenv is a json file create by me
> cat /etc/myenv
{
  "Environment":
  {
    "EXTENSIONS_GALLERY":
    {
      "serviceUrl": "https://test888.com/api",
      "itemUrl": "https://test888.com/item",
      "resourceUrlTemplate": "https://test888.com/files/{publisher}/{name}/{version}/{path}"
    }
  }
}

but after I restart the daemon with the following command:
> sudo systemctl daemon-reload
> sudo systemctl restart code-server@jeff

it seems not work.
Do I make some mistakes ? Or the systemd can't read json format file ?

#Method 2:  I add Environment="EXTENSIONS_GALLERY=<json>" under the [Service] block
it seems not work if I write the following line under the [Service] block

Environment="EXTENSIONS_GALLERY={"serviceUrl":"https://test888.com"}"

Could you give me an example file which can show how to write the format of Environment="EXTENSIONS_GALLERY=<json>" ?

Thank you very much !

**Reply by code-asher** [reply]
Ah yup the environment file is shell syntax, not JSON.  So it would look like this:

/etc/myenv:

```
EXTENSIONS_GALLERY='{"serviceUrl":"https://test888.com"}'
```

If you wanted to use `Environment` then I think this should work (note the placement of quotes, my initial example was a bit off since the value itself contains quotes):
```
Environment=EXTENSIONS_GALLERY='{"serviceUrl":"https://test888.com"}'
```

**Reply by jeffliao888** [reply]
Dear @code-asher , thank you for your help.

I have set the Environment already: 
Environment='EXTENSIONS_GALLERY={"serviceUrl":"https://test888.com"}'

and then I checked with the following command to make sure the Environment is setting in the right place:
> systemctl show code-server@jeff -p Environment
Environment=EXTENSIONS_GALLERY={"serviceUrl":"https://test888.com/"}

then I use code-server IDE with a brower (http://localhost:8080)  and  login already.
then I clicked the "Extensions" icon in the side-bar and choose the "POPULAR", there is an error message show as the following:
"Error while fetching extensions. getaddrinfo ENOTFOUND open-vsx.org"

Do I need to setup anything else so that I can connect to my code-marketplace server(test888.com) ?

Thank you very much !

**Reply by code-asher** [reply]
Hmm strange, that does make it seem like the environment variable is not actually being set correctly somehow but the syntax looks correct to me.

If you open the integrated terminal in code-server and run `printenv EXTENSIONS_GALLERY` what do you see?

**Reply by code-asher** [reply]
Also you probably already did this but just to make sure did you restart the service after making your changes?

**Reply by code-asher** [reply]
Also I am still not sure quoting the way you currently have it will work.  In the shell at least it results in invalid JSON.  You might want to try moving the single quotes to match the second example here:

```console
$ export EXTENSIONS_GALLERY={"serviceUrl":"https://test888.com/"}
$ printenv EXTENSIONS_GALLERY
{serviceUrl:https://test888.com/} # invalid json
$ export EXTENSIONS_GALLERY='{"serviceUrl":"https://test888.com/"}'
$ printenv EXTENSIONS_GALLERY
{"serviceUrl":"https://test888.com/"} # valid json
```

**Reply by jeffliao888** [reply]
Dear @code-asher , thank you for your help.

Is there another easy way to up the code-server so that there is no need to export the Environment value?  Can the docker version work?

Thanks !

**Reply by code-asher** [reply]
Is it still not working for you?

If you were to run Docker you would do it like this:

```
docker run --env EXTENSIONS_GALLERY='{"serviceUrl":"https://test888.com/"}' codercom/code-server
```

Either way you have to provide the environment variable though.

The only other way is to edit `lib/vscode/product.json` and put the extension gallery JSON there.  For example:

```json
{
  // other stuff up here
  "aiConfig": {
    "ariaKey": "code-server"
  },
  "extensionsGallery": {
    "serviceUrl": "https://test888.com"
  }
}
```

**Reply by jeffliao888** [reply, marked as answer]
Dear @code-asher , thank you for your help !

I have a good news to tell you that my code-server can connect to code-marketplace server now.
I have make a mistake to the file /usr/lib/systemd/system/code-server@.service
I made a change from
   Environment='EXTENSIONS_GALLERY={"serviceUrl":"https://test888.com/"}'
to 
   Environment='EXTENSIONS_GALLERY={"serviceUrl":"https://test888.com/api"}'
so that my code-server can connect to my code-marketplace server(test888.com) now.

Thank you very much for your great help !

**Reply by code-asher** [reply]
Ahh the `/api`, that makes sense.  Still strange it was defaulting to Open VSX, it should have just errored instead.

But glad you got it working!
